examples/vision/nerf.py

```diff
@@ -0,0 +1,810 @@
+
+    # Create and return the model.
+    return keras.Model(inputs=inputs, outputs=outputs)
+
+
+def render_rgb_depth(model, rays_flat, t_vals, rand=True, train=True):
+    """
+    Generates the rgb image and depth map from the prediction of the
+    model.
+    Args:
+        - model: The MLP model that is trained to predict the rgb and
+            volume density of the volumetric scene.
+        - rays_flat: The flattened rays that serves as the input to
+            the NeRF model.
+        - t_vals: The sample points for the rays.
+        - rand: Choice to randomise the sampling strategy.
+        - train: Whether the model is in training or testing phase.
+    """
+    # Get the predictions from the nerf model and reshape it.
+    if train:
+        predictions = model(rays_flat)
+    else:
+        predictions = model.predict(rays_flat)
+    predictions = tf.reshape(predictions, shape=(BATCH_SIZE, H, W, NUM_SAMPLES, 4))
+
+    # Slice the predictions into rgb and sigma.
+    rgb = tf.sigmoid(predictions[..., :-1])
+    sigma_a = tf.nn.relu(predictions[..., -1])
+
+    # Get the distance of adjacent intervals.
+    delta = t_vals[..., 1:] - t_vals[..., :-1]
+    # delta shape = (num_samples)
+    if rand:
+        delta = tf.concat(
+            [delta, tf.broadcast_to([1e10], shape=(BATCH_SIZE, H, W, 1))], axis=-1
+        )
+        alpha = 1.0 - tf.exp(-sigma_a * delta)
+    else:
+        delta = tf.concat(
+            [delta, tf.broadcast_to([1e10], shape=(BATCH_SIZE, 1))], axis=-1
+        )
+        alpha = 1.0 - tf.exp(-sigma_a * delta[:, None, None, :])
+
+    # Get transmittance.
+    exp_term = 1.0 - alpha
+    epsilon = 1e-10
+    transmittance = tf.math.cumprod(exp_term + epsilon, axis=-1, exclusive=True)
+    weights = alpha * transmittance
+    rgb = tf.reduce_sum(weights[..., None] * rgb, axis=-2)
+
+    if rand:
+        depth_map = tf.reduce_sum(weights * t_vals, axis=-1)
+    else:
+        depth_map = tf.reduce_sum(weights * t_vals[:, None, None], axis=-1)
+    return (rgb, depth_map)
+
+
+"""
+## Training
+
+The training step is gathered in a `keras.Model` class so that we can
+make use of the `model.fit` call.
+"""
+
+loss_tracker = keras.metrics.Mean(name="loss")
+psnr_metric = keras.metrics.Mean(name="psnr")
+
+
+class NeRF(keras.Model):
+    def __init__(self, nerf_model):
+        super().__init__()
+        self.nerf_model = nerf_model
+
+    def compile(self, optimizer, loss):
+        super().compile()
+        self.optimizer = optimizer
+        self.loss = loss
+
+    def train_step(self, inputs):
+        # Get the images and the rays.
+        (images, rays) = inputs
+        (rays_flat, t_vals) = rays
+
+        with tf.GradientTape() as tape:
+            # Get the predictions from the model.
+            rgb, _ = render_rgb_depth(
+                model=self.nerf_model, rays_flat=rays_flat, t_vals=t_vals, rand=True
+            )
+            loss = self.loss(images, rgb)
```

Comment:
`loss = self.loss(...)` is confusing, prefer using `loss_fn` to make it clear that `self.loss_fn` is a function.

Alternatively you could use

`loss_value = self.loss(...)`